### ISSUE TITLE
perception roadmap

### ISSUE THREAD

**Issue by alemelis** [KojiShiono/Stanley2.0#3]
Here's is a proposal for the steps to be done in the perception node, feel free to modify it

- [x] Obtain training data from `/camera/image_raw`(Artem) 
Images are on google drive shared folder.
- [x] Training data augmentation (if needed)
- [x] Decide wether use a CV or DL approach

We go for DL:
~- [ ] Training data pre-processing~
~- [ ] Define NN architecture~
~- [ ] Select an optimisation and validation strategy~
~- [ ] Training~
~- [ ] Add DL predictor to ROS node~

If we were to go for CV:
- [x] Channel filtering/thresholding to isolate red/green/yellow dots in the upper part of the image
https://github.com/KojiShiono/Stanley2.0/blob/c0c7670308b2c74b642916cbd8033281e575425b/ros/src/tl_detector/tl_detector.py#L200
- [x] Traffic light tracking
- [x] Add CV pipeline to ROS node

Once we are able to find and classify traffic lights in the simulator:
- [x] Link traffic light position to closest waypoint
https://github.com/KojiShiono/Stanley2.0/blob/c0c7670308b2c74b642916cbd8033281e575425b/ros/src/tl_detector/tl_detector.py#L196
- [x] Publish traffic light colour and position

Now it'll be the time to test the classifier with real images (there are two ROS bags for this):
- [x] Extract images from ROS bags https://github.com/KojiShiono/Stanley2.0/blob/6efd3de43197260195d7f252e62a9125fa5f8839/traffic_light_bag_files/bag_extractor.py#L7
- [x] Label (manually) images. There are 4 classes (no light, red, yellow, green) https://github.com/KojiShiono/Stanley2.0/blob/591ad196d4adc11254d091d1d6f678579538138c/traffic_light_bag_files/labeler.py#L2
- [x] Tune classification algorithm

**Comment by alemelis** [comment]
Regarding the CNN architecture, I stumbled upon [this medium post](https://medium.freecodecamp.org/recognizing-traffic-lights-with-deep-learning-23dae23287cc) that we can use for help. I found also an [implementation of SqueezeNet in keras](https://github.com/DT42/squeezenet_demo) that can be useful to start.

I didn't have luck in finding the Nexar traffic light dataset. However, the [LISA Traffic Light dataset](http://cvrr.ucsd.edu/vivachallenge/index.php/traffic-light/traffic-light-detection/) seems to be freely available.
